src/NHibernate/Loader/JoinWalker.cs

```diff
@@ -168,6 +168,7 @@ protected virtual SqlString GetWithClause(string path, string pathAlias)
 			IJoinable joinable = type.GetAssociatedJoinable(Factory);
 
 			string subalias = GenerateTableAlias(associations.Count + 1, path, subPathAlias, joinable);
+			var qc = joinable.IsCollection ? (IQueryableCollection) joinable : null;
```

Comment:
This is safe `IQueryableCollection` casting. As it's anyway applied for all collections inside `OuterJoinableAssociation.IsManyToManyWith`